File: src/effs/crash_point.js
// CrashPoint.stop — JavaScript twin of crash_point.c. Bend owns policy.

function stop(name) {
  try {
    const requested = Buffer.from(io_bytes(name));
    if (process.platform !== "darwin" && process.platform !== "linux") {
      const os = require("os");
      return io_fail(Math.abs(os.constants.errno.ENOTSUP ?? 95));
    }
    const fs = require("fs");
    const marker = Buffer.concat([Buffer.from("crash_point_reached="), requested, Buffer.from("\n")]);
    let offset = 0;
    while (offset < marker.length) {
      const written = fs.writeSync(2, marker, offset, marker.length - offset);
      if (written <= 0) {
        return io_fail(5);
      }
      offset += written;
    }
    process.kill(process.pid, "SIGSTOP");
    return io_done({ $: "Unit" });
  } catch (error) {
    return io_fail(Math.abs(error.errno ?? 5));
  }
}

File: src/effs/crash_point.c
// CrashPoint.stop — irreducible host boundary. Bend decides whether the named
// checkpoint is active; this effect only emits its witness and stops the process.

Term stop_run(Env e, Term* f, IoWork* w) {
  uint64_t length = 0;
  char* name = io_cstr(e, f[0], &length);
  (void)w;

#if defined(__APPLE__) || defined(__linux__)
  const char marker[] = "crash_point_reached=";
  int marker_ok = write(STDERR_FILENO, marker, sizeof(marker) - 1) == sizeof(marker) - 1
    && write(STDERR_FILENO, name, length) == length
    && write(STDERR_FILENO, "\n", 1) == 1;
  free(name);
  if (!marker_ok) {
    return io_fail(e, errno != 0 ? errno : EIO, NULL);
  }
  if (kill(getpid(), SIGSTOP) != 0) {
    return io_fail(e, errno, NULL);
  }
  return io_done(e, term_pak(CID_UNIT, 0));
#else
  free(name);
  return io_fail(e, ENOTSUP, NULL);
#endif
}

static void __attribute__((constructor)) stop_use(void) {
  io_eff(CID_STOP, stop_run, 0);
}
